src/Package/Impl/Publishing/Handlers/RmdPublishHandler.cs

```diff
@@ -29,45 +30,48 @@ internal sealed class RmdPublishHandler : IMarkdownFlavorPublishHandler {
+            long outputRmdBlobId = 0;
 
-            // Perform markdown rendering on the blob
-            string outputFile = Path.GetFileName(outputFilePath);
-            string format = GetDocumentTypeString(publishFormat);
-            string ext = Path.GetExtension(outputFilePath).Substring(1);
-            var publishResult = await session.EvaluateAsync($"rtvs:::rmarkdown_publish(blob_id = {rmdBlobId}, output_filename = '{outputFile}', output_format='{format}', encoding='cp{encoding.CodePage}')", REvaluationKind.Normal);
-            JObject obj = publishResult.Result as JObject;
-            JArray jsonBlobids = obj["blob.ids"] as JArray;
-            JArray jsonFileNames = obj["file.names"] as JArray;
+            try {
+                // Send the markdown to R-Host for processing as a blob
+                rmdBlobId = await session.SendFileAsBlobAsync(fs, inputFilePath);
+                if (rmdBlobId <= 0) {
+                    return;
+                }
 
-            // The response should contain the blob id and the output file name
-            long[] blobids = new long[jsonBlobids.Count];
-            Dictionary<long, string> fileNameMap = new Dictionary<long, string>();
-            int i = 0;
-            foreach (var jblob in jsonBlobids) {
-                blobids[i] = jblob.Value<long>();
-                fileNameMap.Add(blobids[i], jsonFileNames[i].Value<string>());
-                ++i;
-            }
+                // Perform markdown rendering on the blob
+                JObject jsonObj = await RMarkdownRenderAsync(session, rmdBlobId, outputFilePath, publishFormat, encoding);
 
-            // Get the published content from R-Host
-            var data = await session.GetBlobAsync(blobids);
+                // Markdown results should be in a single blob (file) output. The response for markdown rendering should contain 
+                // the blob id and the output file name.
+                JArray jsonBlobids = (JArray)jsonObj["blob_ids"];
+                JArray jsonFileNames = (JArray)jsonObj["file_names"];
+                if (jsonBlobids.Count != 1 || jsonFileNames.Count != 1) {
+                    return;
+                }
 
-            // Save it locally 
-            string outputFolder = Path.GetDirectoryName(outputFilePath);
-            foreach (Blob blob in data) {
-                string fullPath = Path.Combine(outputFolder, fileNameMap[blob.BlobId]);
-                File.WriteAllBytes(fullPath, blob.Data);
+                // Get the published content from R-Host
+                outputRmdBlobId = jsonBlobids[0].Value<long>();
+                string fullOutputFilePath = Path.Combine(Path.GetDirectoryName(outputFilePath), jsonFileNames[0].Value<string>());
+                await session.GetFileFromBlobAsync(fs, outputRmdBlobId, fullOutputFilePath);
+
+            } catch (IOException ex) {
+                coreShell.ShowErrorMessage(ex.Message);
+            } catch (RException) {
```

Comment:
We should probably say something to the user when this happens. This would indicate that the R expression doing the export has failed with an error, which can also happen in all the usual corner cases (I/O failure etc).